# SD-TSIA211 : RECOMMENDATION

## Remi Jaylet

### imports

In [25]:
from movielens_utils import load_movielens, objective
import numpy as np

### 1 - Presentation of the models

In [42]:
R, mask = load_movielens("ml-100k/u.data", minidata=False)
print("R = ", R)
print("mask = ", mask)
print("dimensions = (", len(R), ",", len(R[0]), ")")
print("number of ratings :", sum(sum(mask)))

R =  [[5. 3. 4. ... 0. 0. 0.]
 [4. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [5. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 5. 0. ... 0. 0. 0.]]
mask =  [[ True  True  True ... False False False]
 [ True False False ... False False False]
 [False False False ... False False False]
 ...
 [ True False False ... False False False]
 [False False False ... False False False]
 [False  True False ... False False False]]
dimensions = ( 943 , 1682 )
number of ratings : 100000


In [31]:
R, mask = load_movielens("ml-100k/u.data", minidata=True)
print("dimensions = (", len(R), ",", len(R[0]), ")")

dimensions = ( 100 , 200 )


When the minidata option is activated, the function only returns a small portion of the R matrix of only 100 users over 200 films.

The dimensions of the matric suggest that are ratings from 943 users on 1682 movies.
By calculation the number of True values in the mask, we can we that there is 100000 ratings ranging from 0 to 5.

### 2 - Find P when $ Q^0 $ is fixed

In [57]:
"""import numpy as np

def objective(P, Q0, R, mask, rho):

    tmp = (R - Q0.dot(P)) * mask
    val = np.sum(tmp ** 2)/2. + rho/2. * (np.sum(Q0 ** 2) + np.sum(P ** 2))
    grad_P = (Q0.transpose()).dot(tmp) + rho*P

    return val, grad_P

from scipy.optimize import check_grad

R, mask = load_movielens("ml-100k/u.data")
rho = 0

def func(P,Q0):
    return (P, Q0, R, mask, rho)[0]

def grad(P,Q0):
    return (P, Q0, R, mask, rho)[1]

check_grad(func, grad, [0, 0])
"""
import numpy as np
from scipy.sparse import csc_matrix
from scipy.sparse.linalg import svds, eigs
R = [[1, 0, 1], [5, 1, 2], [0, 1, 0], [0, 0, 3]]
u, s, vt = svds(R, k=2)

ValueError: matrix type must be 'f', 'd', 'F', or 'D'